Pattern library / design system setup

thibaudcolas opened the pull request:
Upgrades our Storybook setup so it’s more usable for any and all UI development, and includes Django templates rendering via [django-pattern-library](https://github.com/torchbox/django-pattern-library) and [storybook-django](https://github.com/torchbox/storybook-django). 

I think this setup is ready to be merged and used, there will be some additional things to do but I’d like to do them further down the line:

- Hosting – having an always-updated site where we can see the pattern library in `main`.
- Better documentation – I’ll rework all our UI dev docs in a separate PR.
- Documentation within Storybook – will be done gradually over multiple UI development pieces.
- Automated tests in CI (see #7176).

There are a few not-very-related changes in here worth mentioning:

- Our ESLint `lint:js` task had `max-warnings` set up, even though I believe LB had removed this quite a few commits ago. Most likely this got added back in a faulty conflicts resolution.
- I forgot to switch one call to `npm run dist` to the `npm run build` command

---

* [x] Do the tests still pass? (https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
* [x] Does the code comply with the style guide? (Run `make lint` from the Wagtail root)
* ~For Python changes: Have you added tests to cover the new/fixed behaviour?~
* [x] For front-end changes: Did you test on all of [Wagtail’s supported environments](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)?
    * Chrome only
* [x] For new features: Has the documentation been updated accordingly?

thibaudcolas commented on client/src/components/Sidebar/Sidebar.stories.tsx:
```diff
@@ -9,8 +9,6 @@ import { LinkMenuItemDefinition } from './menu/LinkMenuItem';
 import { SubMenuItemDefinition } from './menu/SubMenuItem';
 import { initFocusOutline } from '../../utils/focus';
 
-import '../../../../wagtail/admin/static/wagtailadmin/css/sidebar.css';
```
This got moved to the Storybook `preview.js`.

thibaudcolas commented on client/storybook/Welcome.stories.mdx:
```diff
@@ -0,0 +1,7 @@
+
+# Welcome to Wagtail’s design system
+
+Wagtail’s admin interface is built with reusable UI components, structured as a design system for ease of maintenance and third-party reuse.
```
This is just a placeholder, actual content for this "home" screen will be further fleshed out once we have more docs in here (design tokens, an overview of all components, npm package, etc.).

thibaudcolas commented on client/storybook/preview-body.html:
```diff
@@ -0,0 +1,14 @@
+        sprite.innerHTML = html
+        const symbols = Array.from(sprite.querySelectorAll('symbol'));
+        window.WAGTAIL_ICONS = symbols.map((s) => s.id.replace('icon-', ''))
+    });
```
This questionable indentation is due to djhtml – I’ll report an issue there.

thibaudcolas commented on client/storybook/preview.js:
```diff
@@ -2,10 +2,13 @@ import '../tests/stubs';
+window.PATTERN_LIBRARY_API = '/pattern-library/api/v1/render-pattern';
 
 export const parameters = {
-  actions: { argTypesRegex: '^on[A-Z].*' },
```
Wasn’t being used.

thibaudcolas commented on wagtail/admin/templates/wagtailadmin/shared/animated_logo.stories.tsx:
```diff
@@ -0,0 +1,14 @@
+import template from './animated_logo.html';
+
+export default {
+  parameters: {
+    docs: {
+      source: { code: template },
+    },
+  },
```
I’ve added this inconsistently. If people find it useful I’d like to do it systematically with a utils-style function to reduce the boilerplate needed.

thibaudcolas commented on wagtail/tests/settings_ui.py:
```diff
@@ -5,4 +5,19 @@
 
 DATABASES["default"]["NAME"] = "ui_tests.db"  # noqa
 
-WAGTAIL_EXPERIMENTAL_FEATURES = {"slim-sidebar"}
```
This no longer does anything useful since the slim sidebar is now the default.

thibaudcolas commented on wagtail/tests/settings_ui.py:
```diff
@@ -5,4 +5,19 @@
+    # Groups of templates for the pattern library navigation. The keys
+    # are the group titles and the values are lists of template name prefixes that will
+    # be searched to populate the groups.
+    "SECTIONS": (("components", ["wagtailadmin/shared"]),),
```
django-pattern-library requires this being set so it knows which templates it can render – I’ve just gone ahead with this setup for the time being, but there is nothing preventing us from using it for all templates.

thibaudcolas commented on wagtail/tests/urls.py:
```diff
@@ -42,6 +44,16 @@
+        path(
+            "pattern-library/api/v1/sprite",
+            home.sprite,
+            name="pattern_library_sprite",
```
This is needed so we can fetch the sprite with a stable URL.

lb- commented on docs/contributing/developing.rst:
```diff
@@ -282,6 +282,22 @@ This must be done after every change to the source files. To watch the source fi
+    $ ./wagtail/tests/manage.py migrate
+    $ ./wagtail/tests/manage.py createcachetable
+    $ ./wagtail/tests/manage.py runserver 0:8000
+    $ npm run storybook
```
not sure if needed but it might be good to add a note that `npm run storybook` should be run in a different terminal/ssh or similar

Also, it would be good to have here that the default port for the actual viewing of the storybook UI is at `http://localhost:6006/`

thibaudcolas replied:
👍

lb- commented on package.json:
```diff
@@ -109,9 +112,9 @@
+    "fix:js": "eslint --ext .js,.ts,.tsx --fix .",
     "format": "prettier --write '**/?(.)*.{css,scss,js,ts,tsx,json,yaml,yml}'",
-    "lint:js": "eslint --ext .js,.ts,.tsx --report-unused-disable-directives --max-warnings 16 ./client",
+    "lint:js": "eslint --ext .js,.ts,.tsx --report-unused-disable-directives .",
```
Thanks for fixing this, gonna assume that was me sorry :) 

Could you please split out these fixes to a separate commit, just will be easier to look at in the history

lb- commented on tsconfig.json:
```diff
@@ -10,5 +10,17 @@
+    "static",
+    // Files with template syntax.
+    "wagtail/admin/templates/wagtailadmin/edit_handlers/inline_panel.js",
+    "wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotions_formset.js"
```
cannot wait until we can fix these :)

lb- reviewed (changes requested):
I was able to run the storybook locally without too many issues, this looks great.

I have a few comments added, but this looks good to me overall, if you could rebase and clean up the commit history a bit that would be epic.

---

# Other questions for future (not a blocker for this PR)

## 1. Errors for some commenting items 
A note for future though http://localhost:6006/?path=/story/commenting-comment--saving does not work -> Your focus-trap must have at least one container with at least one tabbable node in it at all times (the comment deleting also breaks).

## 2. Alternative to markdown files

Is it possible that the Django Pattern Library can pull in the Django comments to show on the docs page?

E.g. 
wagtail/admin/templates/wagtailadmin/shared/header.html
```
{% comment %}

    Variables accepted by this template:
```

I would think this would be preferable in most cases instead of a standalone markdown file, but maybe I am missing something.

## 3. Folder Structure

Getting into the weeds I know but it is possible we would adopt a bit more of a nested structure for shared templates?

```
wagtail/admin/templates/wagtailadmin/shared
  - header/
    - header.html
    - header.md
    - header.stories.tsx
```

## 4. non-relative imports

Also a nitpick but could we avoid the 
````
import TemplatePattern from '../../../../../client/storybook/TemplatePattern';
```
And instead use something like

```
import { TemplatePattern } from 'storybook';
```

lb- commented:
![screenshot-localhost_6006-2022 02 23-07_36_00](https://user-images.githubusercontent.com/1396140/155223652-3d8918b9-8244-42ef-be51-59a0bca162f4.png)

lb- commented on client/storybook/preview.js:
```diff
@@ -2,10 +2,13 @@ import '../tests/stubs';
 import '../../wagtail/admin/static_src/wagtailadmin/scss/sidebar.scss';
+import './preview.scss';
+
+window.PATTERN_LIBRARY_API = '/pattern-library/api/v1/render-pattern';
```
Is there any way to avoid this being a global window variable, I know it is in the context of storybook only but could create a risk of being written by other code accidentally and adds a bit of indirection.

Maybe it can be an exported const somewhere?

thibaudcolas replied:
I have refactored this to be hard-coded for now (it’s only used in two places). Will try to think of a better way to do this later on, as part of a new release of `storybook-django`.

lb- commented on wagtail/tests/urls.py:
```diff
@@ -42,6 +44,16 @@
+if apps.is_installed("pattern_library"):
+    urlpatterns += [
+        path(
+            "pattern-library/api/v1/sprite",
```
Should this URL be also pulled out into a shared const OR a window variable, similar to `window.PATTERN_LIBRARY_API`?

thibaudcolas replied:
Yep, I’ve done something similar. I think it’ll only be used in one place so have kept it relatively simple.

lb- reviewed (changes requested):
A few more things.

lb- commented:
Just let me know when you want me to review again, thanks @thibaudcolas

lb- commented on MANIFEST.in:
```diff
@@ -3,3 +3,6 @@ graft wagtail
 global-exclude *.py[co]
+global-exclude *.stories.js
+global-exclude *.stories.tsx
+recursive-exclude wagtail/ *.md
```
will we also need to exclude client/ *md ? or will we not used md files for those stories?
Note: there is a md file we may want to keep though `client/README.md`

Maybe client does not get included at all so non-issue (I am still learning the Manifest syntax)

thibaudcolas replied:
Yep, `client` isn’t included at all currently so we should be good.

thibaudcolas commented:
@lb- I’m glad you’re liking it 😄  Hopefully this will make much more sense once there are more components in there. I’ve made the changes you suggested except for some of the high-level points, which I’d like to do as follow-ups to this PR:

> 1. Errors for some commenting items

I have opened #8031 to track this separately.

> 2. Alternative to Markdown files

Yes, that’s definitely possible, if we agree on the logic to pull those comments from the templates. Storybook already assume the docs string is Markdown, so Markdown processing is already there :)

I have updated the header docs to demonstrate how this might work – would like to proceed with this as-is if that’s ok with you, and bring this to the UI team for discussion on what pattern we should use. I can see three different options:

1. Markdown file next to the template / React component
2. "frontmatter" comment (with Markdown syntax optional) inside the template,
3. Documentation written directly inside the `.stories.tsx` file (I have added an example `description` to the `icon`’s `argTypes` definition)

Those three patterns aren’t necessarily incompatible – but we’d want to make sure we agree on which to use when.

> 3. Folder Structure

Yes, I’d like us to use more nesting, but since a lot of those templates will be overridden in third-party sites, I’ve been hesitant to do this from the get-go.

Again I’d like to bring this to the UI team so we have consensus on the folder structure before I make any change.

> 4. non-relative imports

I’d like to keep this as-is for now if that’s ok with you, because I’m expecting `TemplatePattern` will move to be defined inside [storybook-django](https://github.com/torchbox/storybook-django) once I’m happy enough with how it works. So the imports will become `import { TemplatePattern } from 'storybook-django';` like you describe.

lb- commented:
@thibaudcolas thanks for the response to my ideas - I agree 100% with your calls on these, I think that the documentation for the components will probably end up being a mix of things depending on what makes sense but good to know it is possible either way.

Regarding the folder structure - maybe that is not worth the hassle of the change, its only ~3 max per file (story, actual template, markdown) and many will be just two so all good - let's not bombard with too much to decide on.
